src/stream/node/mod.rs

```diff
@@ -249,13 +218,52 @@ where
             return Ok(());
         }
 
+        #[cfg(all(feature = "hardware_counters", target_os = "linux", not(test)))]
```

Comment:
registration should happen in Node::new().

I think we can remove the HardwareMetricGroup now that we operate locally. Also, a lot of code can be removed.

We don't need to reset the group.

for iterators, no need to store the iterator in a variable.
```rust
for counter in self.perf_events.counters.iter() {
}
// instead of
let iterator = self.perf_events.counters.iter();
for counter in iterator {

}
```

The code below has not been verified to compile but should be possible to write it like that.





```rust
// Node constructor
#[cfg(all(feature = "hardware_counters", target_os = "linux", not(test)))]
for counter in self.perf_events.counters.iter() {
    register_histogram!(counter.to_string(),"node" => self.descriptor.clone());
}

// handle message

#[cfg(all(feature = "hardware_counters", target_os = "linux", not(test)))]
let mut (group, counters) = {
    let group = Group::new()?;
    let counters = Vec::with_capacity(self.perf_events.counters.len());
    for hardware_counter in self.perf_events.counters.iter() {
        let counter = Builder::new()
                        .group(&mut hardware_metric_group.group)
                        .kind(val.get_hardware_kind())
                        .build()?;
                        
        counters.push((hardware_counter.to_string(), counter));
    }
    (group, counters)
};

#[cfg(all(feature = "hardware_counters", target_os = "linux", not(test)))]
group.enable()?;

// do work

#[cfg(all(feature = "hardware_counters", target_os = "linux", not(test)))]
group.disable()?;

#[cfg(all(feature = "hardware_counters", target_os = "linux", not(test)))]
let counts = group.read()?;

#[cfg(all(feature = "hardware_counters", target_os = "linux", not(test)))]
for (metric_name, counter) in counters.iter() {
  histogram!(String::from(metric_name), counts[counter] as f64, "node" => self.descriptor.clone());
}
```